draft-irtf-cfrg-opaque.md

```diff
@@ -428,39 +429,52 @@ with the following API:
-We let `Npk` and `Nsk` denote the size of public and private keys, respectively,
-used in the AKE. The AKE protocol must provide the following functions:
+OPAQUE relies on a key recovery mechanism so that a client can engage in the AKE.
+The key recovery mechanism MUST enable deterministic recovery of authentication material
+compatible with the AKE given the OPRF output and additional data.
+This additional data can be encapsulated in a specific `Envelope` structure, saved on
+the server during registration, and provided to the client during login.
 
-- RecoverPublicKey(private_key): Recover the public key related to the input `private_key`.
-- GenerateAuthKeyPair(): Return a randomly generated private and public key pair. This can be
-  implemented by generating a random private key `sk`, then computing `pk = RecoverPublicKey(sk)`.
-- DeriveAuthKeyPair(seed): Derive a private and public authentication key pair
-  deterministically from the input `seed`.
+The key recovery mechanism must provide the two functions allowing for the construction
+of that `Envelope` structure and key recovery from that same `Envelope` structure.
 
-The implementation of DeriveAuthKeyPair is as follows:
+We define an API for these functions, but future specifications are free to adapt their
+signatures according to their needs:
 
-~~~
-DeriveAuthKeyPair(seed)
+- CreateEnvelope(randomized_input, additional_data): build and return an `Envelope` structure
+and the client's public key key.
+- RecoverEnvelope(randomized_input, envelope): recover and return the authentication material
+for the AKE.
+ 
+Any key recovery mechanism used in OPAQUE MUST NOT invalidate the security proof of OPAQUE.
+Moreover, a chosen method must be compatible with the chosen AKE. For example, the key
+recovery mechanism specified in {{key-recovery}} directly recovers a private key from a seed,
+and the cryptographic primitive in the AKE must therefore support such a possibility.
 
-Input:
-- seed, pseudo-random byte sequence used as a seed.
+## Authenticated Key Exchange (AKE) Protocol {#deps-ake}
 
-Output:
-- private_key, a private key.
-- public_key, the associated public key.
+OPAQUE additionally depends on an Authenticated Key Exchange (AKE) protocol. This AKE
+MUST hold within 3 messages and allow the initiating client to input it's private key
+in its second step. 
 
-Steps:
-1. private_key = HashToScalar(seed, dst="OPAQUE-HashToScalar")
-2. public_key = ScalarBaseMult(private_key)
-3. Output (private_key, public_key)
-~~~
+This specification defines one particular AKE based on 3DH; see {{ake-protocol}}.
+3DH assumes a prime-order group as described in {{I-D.irtf-cfrg-voprf, Section 2.1}}.
+We let `Npk` and `Nsk` denote the size of public and private keys, respectively,
+used in the AKE.
 
-HashToScalar(msg, dst) is as specified in {{I-D.irtf-cfrg-voprf, Section 2.1}}.
+The AKE protocol must provide the following functions for the client:
```

Comment:
Maybe we can rewrite this as, "The AKE protocol messages are constructed using the following functions:"